### PR TITLE
std::sync::RwLock docs improvement

### PR DESCRIPTION
**Pull request by lorepozo**
Addresses the `RwLock` part of #29377.
r? @steveklabnik

Added examples, links to types, and a small comparison between RwLock and Mutex.

### CODE REVIEW

File: src/libstd/sync/rwlock.rs [hunk 1]
```diff
@@ -24,19 +24,24 @@ use sys_common::rwlock as sys;
...
 /// particular policy will be used.
 ///
 /// The type parameter `T` represents the data that this lock protects. It is
-/// required that `T` satisfies `Send` to be shared across threads and `Sync` to
-/// allow concurrent access through readers. The RAII guards returned from the
-/// locking methods implement `Deref` (and `DerefMut` for the `write` methods)
-/// to allow access to the contained of the lock.
+/// required that `T` satisfies [`Send`] to be shared across threads and
+/// [`Sync`] to allow concurrent access through readers. The RAII guards
+/// returned from the locking methods implement [`Deref`][] (and [`DerefMut`]
```
Comment (estebank) [inline src/libstd/sync/rwlock.rs]: typo

    [`Deref`][]

**Review by estebank — CHANGES_REQUESTED** [review]

File: src/libstd/sync/rwlock.rs (hunk 1 above)
Comment (lucasem) [reply src/libstd/sync/rwlock.rs]: @estebank It was intentional — without the `[]`, the parentheses are parsed as the markdown syntax for a URL. e.g. in [`sync::mpsc::Receiver`](https://github.com/rust-lang/rust/blob/master/src/libstd/sync/mpsc/mod.rs#L300)

Comment (estebank) [reply src/libstd/sync/rwlock.rs]: TIL

**Review by estebank — APPROVED** [review]

**Pull request merged**